lib/analyzer/image-extractor.ts

```diff
@@ -1,110 +1,181 @@
-      });
-      md5Patterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToBuffer(stream).then((value) => {
-            result.md5[header.name] = md5(value);
-          });
+    layerExtract.on("entry", async (header, stream, next) => {
+      if (
+        header.type === "file" ||
+        header.type === "link" ||
+        header.type === "symlink"
+      ) {
+        const name = `/${header.name}`;
+        let buffer: Buffer | undefined;
+        for (const searchAction of searchActions) {
+          if (minimatch(name, searchAction.pattern, { dot: true })) {
+            if (header.type === "file") {
+              // convert stream to buffer
+              buffer = buffer || (await streamToBuffer(stream));
+              // initialize the files associated products dict
+              result[name] = result[name] || {};
+              // go over the callbacks and assign each product under its callback name
+              for (const callback of searchAction.callbacks) {
+                // already found?
+                if (Reflect.has(result[name], callback.name)) {
+                  debug(`duplicate match ${name} for ${searchAction.pattern}`);
+                }
+                // store product
+                result[name][callback.name] = callback.callback(buffer);
+              }
+            } else {
+              // target is a link or a symlink
+              debug(`${header.type} '${header.name}' -> '${header.linkname}'`);
+            }
+          }
         }
-      });
+      }
       stream.resume(); // auto drain the stream
       next(); // ready for next entry
     });
     layerExtract.on("finish", () => {
       // all layer level entries read
       resolve(result);
     });
-    stream.pipe(layerExtract);
+    layerTarStream.pipe(layerExtract);
   });
 }
 
 /**
- * Extract key files textual content and MD5 sum from the specified TAR file.
+ * Retrieve the products of files content from the specified TAR file.
  * @param imageTarPath path to image file saved in tar format
- * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
- * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
- * @returns manifest file and key files by inner layers
+ * @param searchActions array of pattern, callbacks pairs
+ * @returns array of extracted files products sorted by the reverse order of
+ *  the layers from last to first
  */
-async function extractImageKeyFiles(
+async function extractLayersFromTar(
   imageTarPath: string,
-  txtPatterns: string[],
-  md5Patterns: string[] = [],
-): Promise<ExtractedImage> {
+  searchActions: SearchAction[],
+): Promise<SearchActionProducts[]> {
   return new Promise((resolve) => {
     const imageExtract: Extract = extract();
-    let manifest: string;
-    const layers: Layers = {};
+    const layers: { [key: string]: SearchActionProducts } = {};
```

Comment:
layers is `{[key: string]: [key: string]: { [key: string]: string | Buffer };}`? if so - le'ts rethink about it - feels like overgeneric